src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.Converters.cs

```diff
@@ -122,5 +124,99 @@ internal static bool TryGetDefaultSimpleConverter(Type typeToConvert, [NotNullWh
+
+            return options.TryGetJsonTypeInfoCached(typeToConvert, out JsonTypeInfo? typeInfo)
+                ? typeInfo.Converter
+                : GetConverterForType(typeToConvert, options);
+        }
+
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
+        internal static JsonConverter GetConverterForType(Type typeToConvert, JsonSerializerOptions options)
+        {
+            // Priority 1: Attempt to get custom converter from the Converters list.
+            JsonConverter? converter = options.GetConverterFromList(typeToConvert);
+
+            // Priority 2: Attempt to get converter from [JsonConverter] on the type being converted.
+            if (converter == null)
+            {
+                JsonConverterAttribute? converterAttribute = typeToConvert.GetUniqueCustomAttribute<JsonConverterAttribute>(inherit: false);
+                if (converterAttribute != null)
+                {
+                    converter = GetConverterFromAttribute(converterAttribute, typeToConvert: typeToConvert, memberInfo: null, options);
+                }
+            }
+
+            // Priority 3: Fall back to built-in converters and validate result
+            return options.GetCustomOrBuiltInConverter(typeToConvert, converter);
+        }
+
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
+        private static JsonConverter GetConverterFromAttribute(JsonConverterAttribute converterAttribute, Type typeToConvert, MemberInfo? memberInfo, JsonSerializerOptions options)
+        {
+            JsonConverter? converter;
+
+            Type declaringType = memberInfo?.DeclaringType ?? typeToConvert;
+            Type? converterType = converterAttribute.ConverterType;
+            if (converterType == null)
+            {
+                // Allow the attribute to create the converter.
+                converter = converterAttribute.CreateConverter(typeToConvert);
+                if (converter == null)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_SerializationConverterOnAttributeNotCompatible(declaringType, memberInfo, typeToConvert);
+                }
+            }
+            else
+            {
+                ConstructorInfo? ctor = converterType.GetConstructor(Type.EmptyTypes);
+                if (!typeof(JsonConverter).IsAssignableFrom(converterType) || ctor == null || !ctor.IsPublic)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_SerializationConverterOnAttributeInvalid(declaringType, memberInfo);
+                }
+
+                converter = (JsonConverter)Activator.CreateInstance(converterType)!;
```

Comment:
should we just use `ctor` here?

Reply:
I'm guessing we could, but this is just moving existing code. I wouldn't want to change it now.